openwisp_monitoring/check/classes/iperf.py

```diff
@@ -0,0 +1,205 @@
+
+from .. import settings as app_settings
+from .base import BaseCheck
+
+logger = logging.getLogger(__name__)
+
+Chart = load_model('monitoring', 'Chart')
+Metric = load_model('monitoring', 'Metric')
+Device = load_model('config', 'Device')
+Credentials = load_model('connection', 'Credentials')
+AlertSettings = load_model('monitoring', 'AlertSettings')
+DeviceConnection = load_model('connection', 'DeviceConnection')
+
+
+class Iperf(BaseCheck):
+    def check(self, store=True):
+        device = self.related_object
+        device_connection = self._get_device_connection(device)
+        if not device_connection:
+            logger.warning(
+                f'DeviceConnection is not properly set for "{device}", iperf check skipped!'
+            )
+            return
+        # The DeviceConnection could fail if the management tunnel is down.
+        if not self._connect(device_connection):
+            logger.warning(
+                f'Failed to get a working DeviceConnection for "{device}", iperf check skipped!'
+            )
+            return
+        servers = self._get_iperf_servers(device.organization.id)
+
+        # TCP mode
+        command = f'iperf3 -c {servers[0]} -J'
+        res, exit_code = self._exec_command(device_connection, command)
+        result_tcp = self._get_iperf_result(res, exit_code, device, mode='TCP')
+
+        # UDP mode
+        command = f'iperf3 -c {servers[0]} -u -J'
+        res, exit_code = self._exec_command(device_connection, command)
+        result_udp = self._get_iperf_result(res, exit_code, device, mode='UDP')
+
+        if store:
+            # Store iperf_result field 1 if any mode passes, store 0 when both fails
+            iperf_result = result_tcp['iperf_result'] | result_udp['iperf_result']
+            self.store_result(
+                {**result_tcp, **result_udp, 'iperf_result': iperf_result}
+            )
+        device_connection.disconnect()
+        return {**result_tcp, **result_udp, 'iperf_result': iperf_result}
+
+    def _get_device_connection(self, device):
+        """
+        Returns an active SSH DeviceConnection for a device.
+        """
+        openwrt_ssh = UPDATE_STRATEGIES[0][0]
+        device_connection = DeviceConnection.objects.filter(
+            device_id=device.id,
+            update_strategy=openwrt_ssh,
+            enabled=True,
+            is_working=True,
+        ).first()
+        return device_connection
+
+    def _get_iperf_servers(self, organization_id):
+        """
+        Get iperf test servers
+        """
+        org_servers = app_settings.IPERF_SERVERS.get(str(organization_id))
+        return org_servers
+
+    def _exec_command(self, dc, command):
+        """
+        Executes device command
+        """
+        return dc.connector_instance.exec_command(command, raise_unexpected_exit=False)
+
+    def _connect(self, dc):
+        """
+        Connects device returns its working status
+        """
+        return dc.connect()
```

Comment:
**Nitpick**: Why to use a helper function here?

Reply:
Will remove this :+1:  I'm just testing something :)

Reply:
I used this to make `deviceconnection.connect()` easier to mock inside `test_iperf.py`.